torch/csrc/distributed/autograd/rpc_messages/rpc_with_profiling_req.cpp

```diff
@@ -0,0 +1,151 @@
+#include <torch/csrc/distributed/rpc/utils.h>
+#include <torch/csrc/jit/serialization/pickle.h>
+#include <vector>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+using rpc::RpcCommandBase;
+
+// This constructor is called when creating the RpcWithProfilingReq on the
+// client.
+RpcWithProfilingReq::RpcWithProfilingReq(
+    rpc::worker_id_t fromWorkerId,
+    rpc::MessageType messageType,
+    rpc::Message&& wrappedMessage,
+    const torch::autograd::profiler::ProfilerConfig& profilerConfig)
+    : fromWorkerId_(fromWorkerId),
+      messageType_(messageType),
+      wrappedMessage_(std::move(wrappedMessage)),
+      profilerConfig_(profilerConfig) {
+  tensors_ = wrappedMessage_.tensors();
+  TORCH_INTERNAL_ASSERT(
+      messageType_ == rpc::MessageType::RUN_WITH_PROFILING_REQ,
+      "Incorrect message type");
+  wrappedMessageType_ = wrappedMessage_.type();
+}
+
+// this constructor is only called in fromMessage() which is called in
+// deserializeRequest(). It is called when reconstructing the
+// RpcWithProfilingReq on the remote end.
+RpcWithProfilingReq::RpcWithProfilingReq(
+    rpc::worker_id_t fromWorkerId,
+    rpc::MessageType messageType,
+    std::unique_ptr<rpc::RpcCommandBase> wrappedRpc,
+    rpc::MessageType wrappedMessageType,
+    std::vector<torch::Tensor> tensors,
+    torch::autograd::profiler::ProfilerConfig& profilerConfig)
+    : fromWorkerId_(fromWorkerId),
+      messageType_(messageType),
+      wrappedRpc_(std::move(wrappedRpc)),
+      wrappedMessageType_(wrappedMessageType),
+      tensors_(std::move(tensors)),
+      profilerConfig_(profilerConfig) {
+  TORCH_INTERNAL_ASSERT(wrappedRpc_ != nullptr, "wrappedRpc cant be null");
+}
+
+rpc::MessageType RpcWithProfilingReq::wrappedMessageType() const {
+  return wrappedMessageType_;
+}
+
+void RpcWithProfilingReq::setWrappedRpc(
+    std::unique_ptr<RpcCommandBase> wrappedRpc) {
+  wrappedRpc_ = std::move(wrappedRpc);
+}
+
+rpc::Message RpcWithProfilingReq::toMessageImpl() && {
+  // save the original message ID and type before moving it.
+  auto wrappedMsgId = wrappedMessage_.id();
+  auto wrappedMsgType = wrappedMessage_.type();
+  // destructively move the wrappedMessage and get the payload. Now the payload
+  // of wrappedMessage won't be in a valid state.
+  auto wrappedPayload = std::move(wrappedMessage_).movePayload();
+  // The wrapped payload should not be empty
+  TORCH_INTERNAL_ASSERT(
+      !wrappedPayload.empty(), "Wrapped payload should not be empty.");
+  // Create the ivalues to send over. We need to send the original message type
+  // and id, as well as some profiling metadata.
+  // TODO: send profiling key
+  std::vector<at::IValue> ivalues{wrappedMsgType,
+                                  static_cast<int64_t>(profilerConfig_.state),
+                                  profilerConfig_.report_input_shapes,
+                                  profilerConfig_.profile_memory,
```

Comment:
Does it make sense to let `ProfilerConfigure` to have it's own to/fromIValue(s)? So that, when we expand it in the future, the same RPC code would still work.

Reply:
Yes, I think we should do something like this. I discussed a similar idea with @ilia-cher as well, I think we should make it so RPC code won't have to be updated if profiler config changes. This already holds for `Event`s.